File: twa/app/src/main/java/com/yescirculation/nonblockinglife/AlarmScheduleResult.java
package com.yescirculation.nonblockinglife;

/** Outcome of a single alarm scheduling attempt, with a machine-readable reason on failure. */
final class AlarmScheduleResult {
    static final String REASON_INVALID_TIME = "invalid_time";
    static final String REASON_PAST_TIME = "past_time";
    static final String REASON_NO_CLOCK_APP = "no_clock_app";
    static final String REASON_EXACT_ALARM_PERMISSION_REQUIRED = "permission_required";
    static final String REASON_ALARM_MANAGER_UNAVAILABLE = "alarm_manager_unavailable";

    final boolean ok;
    final String reason;

    private AlarmScheduleResult(boolean ok, String reason) {
        this.ok = ok;
        this.reason = reason;
    }

    static AlarmScheduleResult success() {
        return new AlarmScheduleResult(true, null);
    }

    static AlarmScheduleResult failure(String reason) {
        return new AlarmScheduleResult(false, reason);
    }
}

File: twa/app/src/main/java/com/yescirculation/nonblockinglife/alarm/AlarmScheduleResult.kt
package com.yescirculation.nonblockinglife.alarm

/** Outcome of a single alarm scheduling attempt, with a machine-readable reason on failure.  */
internal class AlarmScheduleResult private constructor(val ok: Boolean, val reason: String?) {
    companion object {
        const val REASON_INVALID_TIME: String = "invalid_time"
        const val REASON_PAST_TIME: String = "past_time"
        const val REASON_NO_CLOCK_APP: String = "no_clock_app"
        const val REASON_EXACT_ALARM_PERMISSION_REQUIRED: String = "permission_required"
        const val REASON_ALARM_MANAGER_UNAVAILABLE: String = "alarm_manager_unavailable"

        fun success(): AlarmScheduleResult {
            return AlarmScheduleResult(true, null)
        }

        fun failure(reason: String?): AlarmScheduleResult {
            return AlarmScheduleResult(false, reason)
        }
    }
}